### ISSUE TITLE
Missing member functions in the REFPROP backends (calc_first_saturation_deriv etc)

### ISSUE THREAD

**Issue by dongkeun-oh** [CoolProp/CoolProp#2013]
### Description

In the REFPROP backends (REPROPMixtureBackend and REFPROPBackend), some member functions, particularly referring saturated properties, aren't implemented yet. However, they are necessary for the ExternalMedia library's binary part (libExternalMedia) to instantiate the CoolProp solver object, and to update (post-update) the thermodynamic states.
So, the REFPROP is not available yet ([see this PR](https://github.com/modelica-3rdparty/ExternalMedia/pull/27)) in that OpenModelica library, and I found that the matter is originated from the following missing functions in the REFPROP backends.
- calc_first_saturation_deriv 
- calc_second_saturation_deriv
- calc_first_two_phase_deriv
- calc_first_two_phase_deriv_splined

### Possible Means to implement
According to the old version (4.2.6), the previous REFPROP solver just exploited similar functionalities commonly implemented in the CPState class. So, I guess there are 3 approaches to go ahead.

1. Newly implementing the functions to the REFPROPMixtureBackend class : this is quite straight forward, but the almost same codes of existing ones will be duplicated, which is an ugly way of development.
 
2. Changing the position of the REFPROP backends to derived classes of the HEOS backends: it depends on the design concept of current configurations, which I didn't understand clearly yet; In my point of view, there are some pitfalls in the different manner of construction between the two backends, and it doesn't look consistent to configure them in a strict hierarchy. 

3. Moving those functions to the parent class (i.e. AbstarctState), and making the parent as a template : in this way, the parent should have the abstracted pointers of saturate properties (i.e. SatL and SatV) like
```
template <typename _Backend>
class AbstractState {
...
shared_ptr<_Backend> SatL, SatV;
...

};
```
, and they will be actualized in the definitions of each backend using so-called CRTP, which means, for instance, something like 
```
class REPRPROPMixtureBackend : public AbstarctState<REPRPROPMixtureBackend> 
```
Moreover, another common construction method might be added up owing to the temporary state instances of "End", and "Liq" in calc_first_two_phase_deriv_splined function.
However, it's the most radical way to modify the configuration a lot.

Do you already have a plan on this matter?
Otherwise, what is the best in the meantime?
If you allow, I would try to implement as you guys will advise after discussion.

### Versions

**CoolProp Version:** CoolProp 6.4.2
**Operating System and Version:** Ubuntu 18.04.5 with gcc 7.5.0, and Windows 10 Msys-64bit with gcc 10.2.0
**Access Method:** ExternalMedia in OpenModelica

### Additional Information

Not yet..

**Comment by ibell** [comment]
I wish it was as simple to do this as you would like it to be.  REFPROP does not have the sort of encapsulation required to do something similar to what you propose with SatL and SatV state shared pointers.  That is just the reality.

The good news is that once the phase equilibrium calculation is done in REFPROP, you will get the saturated liquid and saturated vapor densities and compositions.  In principle that is all you need (along with the low-level routines in REFPROP), to implement the derivatives.  I'd recommend you start with the ``first_saturation_deriv``, as that gives the idea, and is the least challenging to do.

In short, option 1 is the way to go.

**Comment by dongkeun-oh** [comment]
Thank you for the immediate answer! 
Let me go ahead on your advice.

**Comment by jfulem** [comment]
@dongkeun-oh, please, are you working on the missing functions in the REFPROP backends? I'm currently testing the REFPROP backend in the Modelica library as well.

**Comment by dongkeun-oh** [comment]
Well, I guess that I have the first draft for a pull request.  So, there will be an update early in the next week^^.
